LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/DownstreamSender.cs

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace LoRaWan.NetworkServer.BasicsStation
+{
+    using System;
+    using System.IO;
+    using System.Text;
+    using System.Text.Json;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using LoRaTools.LoRaPhysical;
+    using LoRaTools.Regions;
+
+    internal class DownstreamSender : IPacketForwarder
+    {
+        private readonly WebSocketWriterRegistry<StationEui, string> socketWriterRegistry;
+        private readonly IBasicsStationConfigurationService basicsStationConfigurationService;
+
+        public DownstreamSender(WebSocketWriterRegistry<StationEui, string> socketWriterRegistry, IBasicsStationConfigurationService basicsStationConfigurationService)
+        {
+            this.socketWriterRegistry = socketWriterRegistry;
+            this.basicsStationConfigurationService = basicsStationConfigurationService;
+        }
+
+        public async Task SendDownstreamAsync(DownlinkPktFwdMessage message)
+        {
+            if (message is null) throw new ArgumentNullException(nameof(message));
+            if (message.StationEui == default) throw new ArgumentException($"A proper StationEui needs to be set. Received '{message.StationEui}'.");
+
+            if (this.socketWriterRegistry.TryGetHandle(message.StationEui, out var webSocketWriterHandle))
+            {
+                var region = await this.basicsStationConfigurationService.GetRegionAsync(message.StationEui, CancellationToken.None);
+                var payload = Message(message, region);
+                await webSocketWriterHandle.SendAsync(payload, CancellationToken.None);
+            };
+        }
+
+        private static string Message(DownlinkPktFwdMessage message, Region region)
+        {
+            using var ms = new MemoryStream();
+            using var writer = new Utf8JsonWriter(ms);
+
+            writer.WriteStartObject();
+
+            writer.WriteString("msgtype", LnsMessageType.DownlinkMessage.ToBasicStationString());
+            writer.WriteString("DevEui", message.DevEui);
+
+            // 0 is for Class A devices, 2 is for Class C devices
+            // Ideally there Class C downlink frame which answers an uplink which have Tmst and RxDelay set
+            var deviceClassType = message.Txpk.Tmst == 0 && message.LnsRxDelay == 0 ? LoRaDeviceClassType.C : LoRaDeviceClassType.A;
```

Comment:
Help me understand here. The tmst = 0 is the legacy way of identifying class C and LNSrxdelay is the new one right?

Reply:
tmst = 0 was set by internals of our LoRa implementation to immediately send a message.
I am removing it from this check as it would be sufficient to just rely on LnsRxDelay here